### ISSUE TITLE
Unable to setup with create_app(ConfigClass)

### ISSUE THREAD

**Issue by yomajo** [nickjj/flask-db#14]
Hey, Nick. New guy with migrations. However I think it's a legitimate issue.
I have both in .env set:
```
FLASK_APP='main.py'
FLASK_DEBUG='true'
ENVIRONMENT='dev'
PYTHONPATH='.'
```

My tree (stripped):
```
├── alembic.ini
├── app
│   ├── __init__.py
│   ├── static
│   └── ...
├── config
│   ├── app_config.py
│   ├── __init__.py
│   ├── ...
├── main.py
├── migrations
│   ├── env.py
│   ├── __init__.py
│   ├── __pycache__
│   ├── script.py.mako
│   ├── seeds.py
│   └── versions
...
```

`flask db init migrations` ran successfully. `flask db reset` even created tables - but did not create alembic version table (should it, at this point?)

my app_config.py:
```
...
configs = {
  'dev'  : DevelopmentConfig,
  'test' : TestingConfig,
  'prod' : ProductionConfig,}
```
My main.py:
```
...
env = os.environ['ENVIRONMENT']
active_config = configs[env]
app = create_app(active_config)
...
```
in migrations/env.py:
```
from ..main import app


# There's no access to current_app here so we must create our own app.
# app = create_app()
db_uri = app.config["SQLALCHEMY_DATABASE_URI"]
db = app.extensions["sqlalchemy"].db

```
On `flask db migrate check` or migrate revision for that matter I get Import error..
```
    from ..main import create_app
ImportError: attempted relative import with no known parent package

```

Is there an easy way out?

**Comment by nickjj** [comment]
Hi,

What if you change `from ..main import create_app` to be the absolute import path to your create_app function?

**Comment by yomajo** [comment]
Sorry don't understand. Are you referring to imports sorcery such as [this](https://stackoverflow.com/a/67692/11537568)?

could you give an example of what you mean? One file I did not show is where create_app is...
`app/__init__.py`:

```
def create_app(config_class: object):
    app = Flask(__name__)
    app.config.from_object(config_class)
    ...
```

**Comment by nickjj** [comment]
I mean instead of using `..` use an absolute import path like `from myapp.somedir.main import create_app`.

**Comment by yomajo** [comment]
JFC...
`from main import app` works.. Mentally I was fixated on execution path matching that of `env.py`.

Creation on new app instance like below,
```
import os
from app import create_app
from config.app_config import configs

env = os.environ['ENVIRONMENT']
active_config = configs[env]
app = create_app(active_config)

```
also works...

Excuse my stupidity, before I close the issue, could you comment on `flask db migrate check` command? I assume this is something like `git status` but for migrations sync info.

However after I have created version and applied it with `flask db migrate`, when I run `flask db migrate check` I get:

**Comment by nickjj** [comment]
It's all good, no problem.

Internally this library looks at `current_app.import_name` to figure out what the `app` path is. If you have a `create_app` function in an `__init__.py` file that'll result in the path being `app`.

Did you manually change that before trying to see if the default generated `env.py` file worked as is?

**Comment by yomajo** [comment]
No, I didn't. And I had no illusions of default env.py working for me, because my `create_app` takes a config argument.

Moving on... I might have found an issue with `flask db migrate check` or `alembic check` for that matter.

After `flask db migrate` (models in sync with database), alembic check raises error:

```
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.ddl.postgresql] Detected sequence named 'amazondetails_id_seq' as owned by integer column 'amazondetails(id)', assuming SERIAL and omitting
INFO  [alembic.ddl.postgresql] Detected sequence named 'note_id_seq' as owned by integer column 'note(id)', assuming SERIAL and omitting
INFO  [alembic.ddl.postgresql] Detected sequence named 'ebaydetails_id_seq' as owned by integer column 'ebaydetails(id)', assuming SERIAL and omitting
INFO  [alembic.ddl.postgresql] Detected sequence named 'invalidorder_id_seq' as owned by integer column 'invalidorder(id)', assuming SERIAL and omitting
INFO  [alembic.ddl.postgresql] Detected sequence named 'etsydetails_id_seq' as owned by integer column 'etsydetails(id)', assuming SERIAL and omitting
INFO  [alembic.ddl.postgresql] Detected sequence named 'amazonwarehousedetails_id_seq' as owned by integer column 'amazonwarehousedetails(id)', assuming SERIAL and omitting
Traceback (most recent call last):
  File "/home/username/project_name/.venv/bin/alembic", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/username/project_name/.venv/lib/python3.11/site-packages/alembic/config.py", line 590, in main
    CommandLine(prog=prog).main(argv=argv)
  File "/home/username/project_name/.venv/lib/python3.11/site-packages/alembic/config.py", line 584, in main
    self.run_cmd(cfg, options)
  File "/home/username/project_name/.venv/lib/python3.11/site-packages/alembic/config.py", line 561, in run_cmd
    fn(
  File "/home/username/project_name/.venv/lib/python3.11/site-packages/alembic/command.py", line 285, in check
    script_directory.run_env()
  File "/home/username/project_name/.venv/lib/python3.11/site-packages/alembic/script/base.py", line 569, in run_env
    util.load_python_file(self.dir, "env.py")
  File "/home/username/project_name/.venv/lib/python3.11/site-packages/alembic/util/pyfiles.py", line 94, in load_python_file
    module = load_module_py(module_id, path)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/username/project_name/.venv/lib/python3.11/site-packages/alembic/util/pyfiles.py", line 110, in load_module_py
    spec.loader.exec_module(module)  # type: ignore
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/home/username/project_name/migrations/env.py", line 85, in <module>
    run_migrations_online()
  File "/home/username/project_name/migrations/env.py", line 79, in run_migrations_online
    context.run_migrations()
  File "<string>", line 8, in run_migrations
  File "/home/username/project_name/.venv/lib/python3.11/site-packages/alembic/runtime/environment.py", line 865, in run_migrations
    self.get_context().run_migrations(**kw)
  File "/home/username/project_name/.venv/lib/python3.11/site-packages/alembic/runtime/migration.py", line 611, in run_migrations
    for step in self._migrations_fn(heads, self):
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/username/project_name/.venv/lib/python3.11/site-packages/alembic/command.py", line 274, in retrieve_migrations
    revision_context.run_autogenerate(rev, context)
  File "/home/username/project_name/.venv/lib/python3.11/site-packages/alembic/autogenerate/api.py", line 525, in run_autogenerate
    self._run_environment(rev, migration_context, True)
  File "/home/username/project_name/.venv/lib/python3.11/site-packages/alembic/autogenerate/api.py", line 583, in _run_environment
    hook(migration_context, rev, self.generated_revisions)
  File "/home/username/project_name/migrations/env.py", line 60, in process_revision_directives
    if config.cmd_opts.autogenerate:
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'Namespace' object has no attribute 'autogenerate'
```

This morning I wiped out alembic.ini, migrations folder and setup to check this directly with alembic
`alembic init migrations` ...
After some time I of playing with env.py I got it to work and `alembic check` worked well in both scenarios:

- when there's pending migration to be applied
- when db and models are in sync

So the issue is definitely with how `env.py` is constructed from flask-db.

My `env.py`:

```
from logging.config import fileConfig

from alembic import context
from sqlalchemy import engine_from_config
from sqlalchemy import pool

from main import app

# There's no access to current_app here so we must create our own app.
db_uri = app.config["SQLALCHEMY_DATABASE_URI"]
db = app.extensions["sqlalchemy"].db

# Provide access to the values within alembic.ini.
config = context.config

# Sets up Python logging.
fileConfig(config.config_file_name)

# Sets up metadata for autogenerate support,
config.set_main_option("sqlalchemy.url", db_uri)
target_metadata = db.metadata

# Configure anything else you deem important, example:
# my_important_option = config.get_main_option("my_important_option")


def run_migrations_offline():
    """
    Run migrations in 'offline' mode.

    This configures the context with just a URL and not an Engine, though an
    Engine is acceptable here as well. By skipping the Engine creation we
    don't even need a DBAPI to be available.

    Calls to context.execute() here emit the given string to the script output.
    """
    url = config.get_main_option("sqlalchemy.url")
    context.configure(
        url=url,
        target_metadata=target_metadata,
        literal_binds=True,
        dialect_opts={"paramstyle": "named"},
    )

    with context.begin_transaction():
        context.run_migrations()


def run_migrations_online():
    """
    Run migrations in 'online' mode.

    In this scenario we need to create an Engine and associate a connection
    with the context.
    """
    # If you use Alembic revision's --autogenerate flag this function will
    # prevent Alembic from creating an empty migration file if nothing changed.
    # Source: https://alembic.sqlalchemy.org/en/latest/cookbook.html
    def process_revision_directives(context, revision, directives):
        if config.cmd_opts.autogenerate:
            script = directives[0]
            if script.upgrade_ops.is_empty():
                directives[:] = []

    connectable = engine_from_config(
        config.get_section(config.config_ini_section),
        prefix="sqlalchemy.",
        poolclass=pool.NullPool,
    )

    with connectable.connect() as connection:
        context.configure(
            connection=connection,
            target_metadata=target_metadata,
            process_revision_directives=process_revision_directives
        )

        with context.begin_transaction():
            context.run_migrations()


if context.is_offline_mode():
    run_migrations_offline()
else:
    run_migrations_online()

```
Basically the only difference is instead of `from app import create_app` I have `from main import app`

Note, that when revision is not applied (database is lagging the models) `alembic check` / `flask db migrate check` produces normal output:
```
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
ERROR [alembic.util.messaging] Target database is not up to date.
  FAILED: Target database is not up to date.

```

**Comment by yomajo** [comment]
Reopening in case no notification was triggered with comment on closed issue.
